src/vm/edge_shape.rs

```diff
@@ -41,11 +41,72 @@ use crate::util::{Address, ObjectReference};
+    /// -   If, in a VM, the data type a slot can hold is a union of references and non-reference
+    ///     values, and the slot is currently holding a non-reference value, such as a small
+    ///     integer, a floating-point number, or any special value such as `true`, `false` or `nil`
+    ///     that do not point to any object, the slot is considered not holding an reference.  This
+    ///     method should return `ObjectReference::NULL` in such cases.
+    ///
+    /// If the slot holds an object reference with tag bits, the returned value shall be the object
+    /// reference with the tag bits removed.
     fn load(&self) -> ObjectReference;
 
     /// Store the object reference `object` into the edge.
+    ///
+    /// FIXME: Currently the subsuming write barrier (`Barrier::object_reference_write`) calls this
+    /// method to perform the actual store to the field.  It only works if the VM does not store
+    /// tag bits in the slot.
+    ///
+    /// FIXME: If the slot contains tag bits, consider overriding the `update_for_forwarding`
+    /// method. See: https://github.com/mmtk/mmtk-core/issues/1033
     fn store(&self, object: ObjectReference);
 
+    /// Update the slot for forwarding.
+    ///
+    /// If the slot is holding an object reference, this method shall call `updater` with that
+    /// reference; if the slot is not holding an object reference, including when holding a NULL
+    /// pointer or holding small integers or special non-reference values such as `true`, `false`
+    /// or `nil`, this method does not need to take further action.  In no circumstance should this
+    /// method pass `ObjectReference::NULL` to `updater`.
+    ///
+    /// If the returned value of the `updater` closure is not `ObjectReference::NULL`, it will be
+    /// the forwarded object reference of the original object, and this method shall update the
+    /// slot to point to the new location; if the returned value is `ObjectReference::NULL`, this
+    /// method does not need to take further action.
+    ///
+    /// This method is called to trace the object pointed by this slot, and forward the reference
+    /// in the slot.  To implement this semantics, the VM should usually preserve the tag bits if
+    /// it uses tagged pointers to indicate whether the slot holds an object reference or
+    /// non-reference values.
+    ///
+    /// The default implementation calls `self.load()` to load the object reference, and calls
+    /// `self.store` to store the updated reference.  VMs that use tagged pointers should override
+    /// this method to preserve the tag bits between the load and store.
+    fn update_for_forwarding<F>(&self, updater: F)
+    where
+        F: FnOnce(ObjectReference) -> ObjectReference,
+    {
+        let object = self.load();
+        if object.is_null() {
+            return;
+        }
+        let new_object = updater(object);
+        if new_object.is_null() {
```

Comment:
Can you do `new_object.is_null() && object == new_object`?

Reply:
I am expecting the `updater` (usually implemented by `trace_object`) to do that test inside (if necessary at all) so that we don't need to check `object == new_object` here.  One example is the nursery (for GenCopy and GenImmix).  It always moves the object, so it doesn't need to check.  It actually does `debug_assert!(!self.plan.is_object_in_nursery(new_object));` which implies `new_object` cannot be equal to `object`.

Reply:
Hm? I don't follow. `trace_object` can't store the new object (if any) into the slot. It will only return an object reference

Reply:
That's right.  `trace_object` doesn't do the storing.  However, the contract of the `updater` closure is, if it returns `ObjectReference::NULL`, then `Edge::update_for_forwarding` will not do the storing.

`process_edge` implements the `updater` closure.  After executing `let new_object = trace_object(object)`, it should check if `new_object == object` and, if they are equal, return `ObjectReference::NULL` from the `updater` closure.  Currently `ProcessEdgesWork::process_edge` is not doing this because I don't want to change the semantics of the existing `ProcessEdgesWork::process_edge`.  We may give it a try by adding `if new_object == object { return ObjectReference::NULL; } else { return new_object; }` there.  IIRC, @qinsoon once tried doing this check after `trace_object`, and it is not always profitable.  We may give it another try, but at this moment, I want to make sure the closure itself doesn't introduce extra cost.

Reply:
I believe Yi's work made it an `Option<ObjectReference>`. That has more overhead since it doesn't condense into a `usize` (given `ObjectReference::NULL` exists). I don't think we should retain the current semantics since we know the current semantics are broken. There is no need to store the same object again into the slot

Reply:
Related issues:
-   https://github.com/mmtk/mmtk-core/issues/574
-   https://github.com/mmtk/mmtk-core/issues/559#issuecomment-1128928487

My previous experiment showed that the performance impact is negligible.  So we have reasons to replace the constant `OVERWRITE_REFERENCE` and the invocation `P::may_move_objects::<KIND>()` with an actual `new_object == object` check. But we'll probably do it in a separate PR to address https://github.com/mmtk/mmtk-core/issues/574 specifically.

Reply:
Sure. Alternatively, we can make a PR for fixing that right now (since it should be simple to fix) and then merge that first, given this PR is a "design" PR and I'm not sure if we've decided what's the best way to review them yet

Reply:
> Sure. Alternatively, we can make a PR for fixing that right now (since it should be simple to fix) and then merge that first

Yes. That's a good idea.

> given this PR is a "design" PR and I'm not sure if we've decided what's the best way to review them yet

I think this can be an "MEP", too, but this may not be as controversal as removing `ObjectReference::NULL`.  Since I already made this PR, I think I will test the performance with the OpenJDK binding and, if it performas well, I'll write up an "MEP" for this, too, and welcome everyone to discuss.  Actually I still have some design questions about this.  One is whether we need a way to update slots atomically.  The LXR branch sometimes does "compare exchange" on slots.